### DISCUSSION TITLE
Integrate JSBSim to AirSim

### DISCUSSION THREAD

**Discussion by jonyMarino** [JSBSim-Team/jsbsim#496] [Q&A]
Hi! 
I am trying to integrate JSBSim as a C ++ library into AirSim as a proof of concept.
We have managed to run a JSBSim model inside AirSim. Now we have the following questions to continue:

First: For the dynamics of the collisions and the sensors, we have to obtain all the kinematics states from jsbsim (position (NED), orientation (Quaternion), linear and angular velocity, and linear and angular acceleration)  in the SI system. By what JSBSim properties are they obtained?

Second: Collisions are calculated within AirSim, so we must set the calculated kinematics states in the running jsbsim model after a collision. How can we do that?

@seanmcleod

**Comment by seanmcleod** [comment]
Here are a couple of property references to get you started for your first question. Note that internally JSBSim doesn't use SI units.

https://github.com/JSBSim-Team/jsbsim/blob/08d2f857b407b5501fe8bba5c245c1051c606c60/src/models/FGPropagate.cpp#L822-L832

https://github.com/JSBSim-Team/jsbsim/blob/08d2f857b407b5501fe8bba5c245c1051c606c60/src/models/FGPropagate.cpp#L873-L875

https://github.com/JSBSim-Team/jsbsim/blob/08d2f857b407b5501fe8bba5c245c1051c606c60/src/models/FGAccelerations.cpp#L348-L354

**Reply by jonyMarino** [reply]
Thanks a lot @seanmcleod! 
It seems to be working. I don't see any of those being the position. We are currently calculating it through position / lat-geod-deg, long-gc-deg and h-sl-meters, like this:

```
double latitude = 111320 * model.GetPropertyValue("position/lat-geod-deg");
            double longitude = 40075000 * model.GetPropertyValue("position/long-gc-deg") * cos(model.GetPropertyValue("position/lat-geod-deg") * (M_PI / 180.0)) / 360;
            double altitude = model.GetPropertyValue("position/h-sl-meters");
```

Now to set the kinematics to JSBSim I'm thinking to restart the model and set the values as initial conditions. Is that possible? How?
I know there may be states that we are losing in the middle, but is there any other way?

**Comment by seanmcleod** [comment]
Yep that was just to get you started and point you to the relevant files as opposed to being a comprehensive list. There are a number of position options with different frames of reference, e.g. ECEF, ECI etc.

In terms of collisions that's potentially going to be tricky. JSBSim has the concept of contact points which can be defined in the aircraft model, e.g. the nose, wing tips, tail etc. which are used for collision detection typically with a ground plane. It also includes details for the landing gear with spring coefficients etc. to model collisions/landing between the gear and the ground.

Now it sounds like you want to use Unreal to perform collision detection and then take some data from it's collision calculations to reset the kinematic states in JSBSim?

**Reply by jonyMarino** [reply]
Thank you very much for following @seanmcleod.
Yes, you are correct, that is what we want to do.

**Comment by seanmcleod** [comment]
@jonyMarino are you planning on implementing all the necessary calculations of the gear forces, reactions etc. as opposed to providing a ground callback to JSBSim for JSBSim to do that?

> I'm thinking to restart the model and set the values as initial conditions. Is that possible? How?

If you do want to go down that route then take a look at the following discussion. 

[Stable outputs when re-running initial conditions](https://github.com/JSBSim-Team/jsbsim/discussions/455)

**Comment by seanmcleod** [comment]
@jonyMarino thinking about this a bit more what are the main reasons for using Unreal to perform the collision detection and the extra complication of trying to then integrate that with JSBSim? As opposed to letting JSBSim perform the collision detection, handle the physics of the collision etc. simply by providing a ground callback to JSBSim?

Otherwise it sounds like a lot of extra work. You would need to grab the `<contact>` data for all the defined contact points of the aircraft model so that Unreal could detect for example the left wingtip impacting the ground say before any of the landing gears, interacting with JSBSim's contact physics for the landing gears etc.

Some examples of the `<contact>` elements that can be defined in a JSBSim model.

https://github.com/JSBSim-Team/jsbsim/blob/08d2f857b407b5501fe8bba5c245c1051c606c60/aircraft/c172p/c172p.xml#L122-L136

https://github.com/JSBSim-Team/jsbsim/blob/08d2f857b407b5501fe8bba5c245c1051c606c60/aircraft/c172p/c172p.xml#L176-L187

I've used the ground callback mechanism in the past with a simulator I built that integrates Prepar3d and JSBSim, i.e. passing the terrain data that I retrieve from Prepar3d for each simulation time-step to JSBSim.

For example for someone like @AOS55 who is wanting to use the combination of AirSim and JSBSim for reinforcement learning with computer vision processing the AirSim scenery and a control loop to then land the aircraft I'm guessing he'd want JSBSim's ground contact processing of the landing gear etc. in order to work out whether the ML agent successfully landed the aircraft or not.

**Reply by jonyMarino** [reply]
Hi @seanmcleod, thanks for the suggestion.

I am willing to evaluate both.
The idea is not to read the JSBSim configuration file but to have a realistic mesh and get the collision from there. We are currently resetting the model with the new position and speed calculated in AirSim. It is not very precise but I think it is enough for what the users need. We also detect if the vehicle landed by its position relative to the ground.
To detect whether the vehicle landed or not, it may be that using JSBSim is better.

According to your suggestion, the contact configuration is a single point, shouldn't it be an area? What would be the C ++ code to apply some force at a certain point? Does anything change for the landing gear?

**Reply by seanmcleod** [reply]
In general a mesh would be overkill right? Aircraft models in JSBSim don't come with a mesh. So one would have to be supplied. 

Here is the code for ground contacts, collisions and gear handling. 

https://github.com/JSBSim-Team/jsbsim/blob/master/src/models/FGLGear.cpp

https://github.com/JSBSim-Team/jsbsim/blob/master/src/models/FGGroundReactions.cpp

**Reply by jonyMarino** [reply]
@seanmcleod We were able to run different JSBSim scripts within AirSim. We handle collisions by setting the initial conditions and applying them with RunIC (). But something we notice is changing the value of other parameters also change in the simulation. Is it possible to directly set all the kinematics values calculated post-collision into JSBSim properties?
The branch we are working on is https://github.com/jonyMarino/AirSim/tree/JSB-AirSim.

**Reply by seanmcleod** [reply]
> But something we notice is changing the value of other parameters also change in the simulation.

Not sure exactly what you're trying to say. Can you rephrase and give more detail?

But I'm guessing you may be referring to something along these lines - https://github.com/JSBSim-Team/jsbsim/discussions/455 ?

**Reply by jonyMarino** [reply]
@seanmcleod  Sorry, and thanks for continuing to answer. We observed that when setting the parameter corresponding to the altitude of the aircraft (position / h-sl-meters), it changed its altitude in runtime. Therefore we wonder if it will be possible to change the position, the orientation, the velocities and the accelerations, all in runtime through the corresponding parameters. This would help us update the JSBSim model after a collision.
We already tried with the initial conditions but they do not have accelerations.

**Comment by Jaeyoung-Lim** [comment]
@jonyMarino You could also try to integrate directly into jsbsim bridge: https://github.com/Auterion/px4-jsbsim-bridge and add a interface for airlib (if this helps)

**Reply by jonyMarino** [reply]
Thanks, @Jaeyoung-Lim. That might be a good idea. The bridge runs JSBSim internally, but we want to run JSBSim inside AirSim. To use it, we would need to compile it into AirSim (which I find difficult), or remove the JSBSim instance from the bridge and exchange the messages with JSBSim through airlib. Do you see that possible?

**Reply by Jaeyoung-Lim** [reply]
> we would need to compile it into AirSim (which I find difficult)

@jonyMarino Not sure if I agree, the jsbsim_bridge does not compile jsbsim into its bridge but links to the library. You can use the same API as is used in the jsbsim bridge. That way you have direct access to jsbsim related states.

**Reply by jonyMarino** [reply]
@Jaeyoung-Lim That's what I meant. The bridge instantiates the jsbsim model inside it. In this way, we should modify it to read and set parameters of the model and at a very high frequency (that of physics). It seems better to me to modify it so that the controller interacts with an external jsbsim instance and at a lower frequency (that of the controller)

**Reply by Jaeyoung-Lim** [reply]
@jonyMarino Hmm interesting, but in this case how would you do a lockstep between airsim and jsbsim? The intention of running jsbsim instance inside the loop was to have full control of the stepping of the simulation (e.g. be able to run jsbsim faster than realtime in the CI)

**Reply by jonyMarino** [reply]
@Jaeyoung-Lim  That's what I wanted to ask you 😆 . We can get the simulation time on the bridge [like here] (https://github.com/Auterion/px4-jsbsim-bridge/blob/53151f341c55b24e47c7783d34bcb0a70dd3a4da/src/jsbsim_bridge.cpp#L172). Wouldn't that be enough?
I think then there is a timeout restraint. In case it is presented to us, I do not know how we should proceed.
Also, AirSim has a continueForTime() API; I don't know if we could use that together with some modification to the bridge to get lockstep.

**Reply by Jaeyoung-Lim** [reply]
@jonyMarino How would you use the simulation time for lockstepping? If I understand correctly, the jsbsim instance is freewheeling outside, you cannot get into to sync with the simulation and certainly not speed the instance up right?

**Reply by jonyMarino** [reply]
@Jaeyoung-Lim JSBSim runs inside our physics thread. So it could be stopped, sped up, or slowed down.
> How would you use the simulation time for lockstepping?  

 You are right. It is not lockstep, but I understand that it would be synchronized since if the simulator takes more or less, the PX4 will adjust accordingly. Am I right?

**Reply by Jaeyoung-Lim** [reply]
> How would you use the simulation time for lockstepping?
> You are right. It is not lockstep, but I understand that it would be synchronized since if the simulator takes more or less, the PX4 will adjust accordingly. Am I right?

@jonyMarino If you are not running on lockstep, the PX4 SITL instance will not adjust the timesteps but just freewheel on it's one. This will result in inconsistent results depending on the system you are running with

**Reply by jonyMarino** [reply]
@Jaeyoung-Lim So what is simulation time used for inside the bridge?
I understand that there must be some way to generate lockstep. If not, how do you do with the gazebo bridge?

**Reply by Jaeyoung-Lim** [reply]
@jonyMarino The simulation time is used to set the timestamps for the HIL_SENSOR messages, which are used for lockstep

> I understand that there must be some way to generate lockstep. If not, how do you do with the gazebo bridge?

Not sure if I understand what you mean. What I was trying to explain was that having jsbsim run as an internal object was a design choice to have good lockstep performance. You can achieve this in other ways depending on what interface you use. However, the simulation should never run on its own to get consistent performance with the PX4 instance (meaning, that if the timestep is 250Hz in the simulation and PX4 should be going back and forth at 250Hz).

> If not, how do you do with the gazebo bridge?

We explicitly control the update so that we run gazebo simulation in lockstep.

**Comment by jonyMarino** [comment]
> We explicitly control the update so that we run gazebo simulation in lockstep.

@Jaeyoung-Lim Then we should do something more like that bridge. Thank you for all the info.
